### PR TITLE
[TypeDeclaration] Use @return type on Generator on ReturnTypeDeclarationRector

### LINKED ISSUE #6989: ReturnTypeDeclarationRector is converting Generator to Iterator
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-master      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.org/demo/1ec89bdc-a24f-68ce-b1af-894c7ad7ac39

```php
<?php

final class DemoFile
{
    /**
     * @return Generator<int>
     **/
    public function generator()
    {
        yield 5;
    }
}
```

### Responsible rules

* `ReturnTypeDeclarationRector`

## Expected Behavior

I was expecting the Rector to add `Generator` as return type

```diff
<?php

final class DemoFile
{
    /**
     * @return Generator<int>
     **/
-    public function generator()
+    public function generator(): Generator
    {
        yield 5;
    }
}
```

But for some reason it prefers `Iterator`

```diff
<?php

final class DemoFile
{
    /**
     * @return Generator<int>
     **/
-    public function generator()
+    public function generator(): \Iterator
    {
        yield 5;
    }
}
```

### PR DESCRIPTION
**Pull request by samsonasik**
Closes https://github.com/rectorphp/rector-src/pull/1792 
Fixes https://github.com/rectorphp/rector/issues/6989

### CODE REVIEW

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules/TypeDeclaration/TypeInferer/ReturnTypeInferer/YieldNodesReturnTypeInfererTypeInferer.php [hunk 1]
```diff
@@ -51,19 +60,44 @@ public function inferFunctionLike(FunctionLike $functionLike): Type
...
+                return self::CLASS_NAME;
+            }
+
+            $typeName = ltrim($returnTagValueNode->type->type->name, '\\');
+            if ($typeName === 'Generator') {
+                return 'Generator';
+            }
+        }
+
+        return self::CLASS_NAME;
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/TypeInferer/ReturnTypeInferer/YieldNodesReturnTypeInfererTypeInferer.php]: This looks like hardcoded resolution from the docblock. What if there is wrong type in `@return`?

File: rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/preserve-generator.php.inc [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /**
+     * @return \Generator<int>
+     **/
+    public function generator(): \Generator
+    {
+        yield 5;
+    }
+}
+?>
```
Comment (TomasVotruba) [inline rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/preserve-generator.php.inc:31]: There should be one more fixture to add return type without any docblock.

File: rules/TypeDeclaration/TypeInferer/ReturnTypeInferer/YieldNodesReturnTypeInfererTypeInferer.php (hunk 1 above)
Comment (samsonasik) [reply rules/TypeDeclaration/TypeInferer/ReturnTypeInferer/YieldNodesReturnTypeInfererTypeInferer.php]: wrong type in `@return` will fallback to `Iterator` as constant `self::CLASS_NAME` defined.

File: rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/preserve-generator.php.inc (hunk 2 above)
Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/preserve-generator.php.inc:31]: There are already fixture for that:

- https://github.com/rectorphp/rector-src/blob/main/rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/generator.php.inc

- https://github.com/rectorphp/rector-src/blob/main/rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/skip_anonymous_function_nested_type.php.inc

- https://github.com/rectorphp/rector-src/blob/main/rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/skip_iterable_array_iterator_co_type.php.inc

- https://github.com/rectorphp/rector-src/blob/main/rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/skip_iterator_already.php.inc

File: rules/TypeDeclaration/TypeInferer/ReturnTypeInferer/YieldNodesReturnTypeInfererTypeInferer.php (hunk 1 above)
Comment (samsonasik) [reply rules/TypeDeclaration/TypeInferer/ReturnTypeInferer/YieldNodesReturnTypeInfererTypeInferer.php]: I added new fixture for wrong type generator via `@return`, it goes to `Iterator` return type https://github.com/rectorphp/rector-src/pull/1794/commits/d18bd5a2089e1c5f4abf76a811016940fd11800e

File: rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/wrong_type_generator.php.inc [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+final class WrongTypeGenerator
+{
+    /**
+     * @return \wrong
+     **/
+    public function generator(): \Iterator
+    {
+        yield 5;
+    }
```
Comment (TomasVotruba) [inline rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/wrong_type_generator.php.inc]: It looks like Rector handles this incorrectly. It should be`\Generator`.

https://www.php.net/manual/en/language.generators.syntax.php

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/wrong_type_generator.php.inc]: Ok, implemented to return Generator when `$functionLike->getReturnType()` is null https://github.com/rectorphp/rector-src/pull/1794/commits/b1affaee2eaceecd2760f703790b657d11b2c089

**Comment by samsonasik** [comment]
It seems make return `Generator` cause error on `AddArrayReturnDocTypeRector`

```bash
There were 3 failures:

1) Rector\Tests\TypeDeclaration\Rector\ClassMethod\AddArrayReturnDocTypeRector\AddArrayReturnDocTypeRectorTest::test with data set #14 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/Fixture/yield_strings.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 class YieldStrings
 {
     /**
-     * @return \Iterator<string>
+     * @return mixed[]
      */
     public function getValues(): iterable
     {

/home/runner/work/rector-src/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:154
/home/runner/work/rector-src/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:115
/home/runner/work/rector-src/rector-src/rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/AddArrayReturnDocTypeRectorTest.php:18

2) Rector\Tests\TypeDeclaration\Rector\ClassMethod\AddArrayReturnDocTypeRector\AddArrayReturnDocTypeRectorTest::test with data set #41 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/Fixture/unique_return_iterator.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 
 final class UniqueReturnIterator
 {
-    /**
-     * @return \Iterator<\Rector\FileSystemRector\ValueObject\AddedFileWithContent[]|\Symplify\SmartFileSystem\SmartFileInfo[]>
-     */
     public function provideData(): Iterator
     {
         $smartFileSystem = new SmartFileSystem();

/home/runner/work/rector-src/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:154
/home/runner/work/rector-src/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:115
/home/runner/work/rector-src/rector-src/rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/AddArrayReturnDocTypeRectorTest.php:18

3) Rector\Tests\TypeDeclaration\Rector\ClassMethod\AddArrayReturnDocTypeRector\AddArrayReturnDocTypeRectorTest::test with data set #54 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/Fixture/return_yield_iterator.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 final class ReturnYieldIterator
 {
     /**
-     * @return \Iterator<string[]>
+     * @return \Generator<string[]>
      */
     public function someMethod()
     {
```

**Comment by samsonasik** [comment]
Make direct return Generator type will cause invalid result on Iterator co type:

```diff
 
 class SkipIterableArrayIteratorCoType
 {
-    public function run(): Iterator
+    public function run(): \Generator
     {
         yield 5;
     }
 
-    public function run2(): Traversable
+    public function run2(): \Generator
     {
         yield 5;
     }
```

**Comment by TomasVotruba** [comment]
> Make direct return Generator type will cause invalid result on Iterator co type:

This should add generator just in case it's empty IMO. In these cases it should skip it.

**Comment by samsonasik** [comment]
The type result is somehow shared with rule `AddArrayReturnDocTypeRector`, but the result is removing/adding each other tests with return type and/or docblock return exists check.

**Comment by samsonasik** [comment]
Fixed 🎉 , modified for both : 

- AddArrayReturnDocTypeRector
- ReturnTypeDeclarationRector

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by samsonasik** [comment]
I am merging it ;)

**Comment by TomasVotruba** [comment]
Thank you :+1: 

Next time please wait a few days before merge. Now I'm not available every day to review on Github and I'd like to merge it knowingly. Thanks

**Pull request merged**